charms/knative-operator/tests/unit/test_charm.py

```diff
@@ -1,67 +1,121 @@
-        expected_plan = {
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {"thing": "🎁"},
-                }
-            },
-        }
-        # Get the httpbin container from the model
-        container = self.harness.model.unit.get_container("httpbin")
-        # Emit the PebbleReadyEvent carrying the httpbin container
-        self.harness.charm.on.httpbin_pebble_ready.emit(container)
-        # Get the plan now we've run PebbleReady
-        updated_plan = self.harness.get_container_pebble_plan("httpbin").to_dict()
-        # Check we've got the plan we expected
-        self.assertEqual(expected_plan, updated_plan)
-        # Check the service was started
-        service = self.harness.model.unit.get_container("httpbin").get_service("httpbin")
-        self.assertTrue(service.is_running())
-        # Ensure we set an ActiveStatus with no message
-        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
+from charm import KnativeOperatorCharm
+
+
+class _FakeResponse:
+    """Used to fake an httpx response during testing only."""
+
+    def __init__(self, code):
+        self.code = code
+
+    def json(self):
+        return {"apiVersion": 1, "code": self.code, "message": "broken"}
+
+
+class _FakeApiError(ApiError):
+    """Used to simulate an ApiError during testing."""
+
+    def __init__(self, code=400):
+        super().__init__(response=_FakeResponse(code))
+
+
+@pytest.fixture
+def harness():
+    """Returns a harnessed charm with leader == True."""
+    harness = Harness(KnativeOperatorCharm)
+    harness.set_leader(True)
+    return harness
+
+
+@pytest.fixture()
+def mocked_codecs(mocker):
+    """Yields a mocked lightkube codecs."""
+    yield mocker.patch("charm.codecs")
+
+
+@pytest.fixture()
+def mocked_client(mocker):
+    """Yields a mocked lightkube Client."""
+    mocked_lightkube_client = MagicMock()
+    mocked_lightkube_client_factory = mocker.patch("charm.Client")
+    mocked_lightkube_client_factory.return_value = mocked_lightkube_client
+    yield mocked_lightkube_client
+
+
+def test_events(harness, mocked_client, mocker):
+    harness.begin()
+    main = mocker.patch("charm.KnativeOperatorCharm._main")
+    pebble_ready = mocker.patch("charm.KnativeOperatorCharm._on_knative_operator_pebble_ready")
+
+    harness.charm.on.install.emit()
+    main.assert_called_once()
+    main.reset_mock()
+
+    harness.charm.on.config_changed.emit()
+    main.assert_called_once()
+    main.reset_mock()
+
+    harness.charm.on.knative_operator_pebble_ready.emit("knative-operator")
+    pebble_ready.assert_called_once()
+    pebble_ready.reset_mock()
+
+
+def test_apply_all_resources(harness, mocked_client, mocked_codecs, mocker):
```

Comment:
I think this test makes more sense as 2 separate ones, one for success and one for failure.

Reply:
Done